lib/clustering/mean_shift.cpp

```diff
@@ -32,67 +32,54 @@ std::vector<Cluster> MeanShift::operator()(const PointCloud& cloud) const {
 	}
 
 	// Initialize some stuff used in the MeanShift loop
-	std::vector<Eigen::VectorXd> prev = points;
-	std::vector<Eigen::VectorXd> curr(points.size());
-	bool has_converged = false;
-  int iterations = 0; //for logging
-
-	// The MeanShift loop
-	while (!has_converged) {
-		iterations++;
-		if (iterations > MAX_ITERATIONS) {
-			BOOST_LOG_TRIVIAL(error) << "MeanShift: max number of iterations exceeded without convergence";
-			break;
-		}
-		// perform 1 iteration on each mode
-		for (int i=0; i<prev.size(); i++) {
-			curr[i] = iterate_mode(prev[i], prev);
-		}
-
-		// Merge step
-		// For every mode..
-		for (int i=0; i<curr.size(); i++) {
-			// Check modes we haven't already executed the (i)-loop for
-			for (int j=i+1; j<curr.size(); j++) {
-				Eigen::VectorXd diff = curr[i] - curr[j];
-
-				if (diff.norm() < MERGE_THRESHOLD) {
-					// Merge clusters. Erase one of the modes corresponding to the clusters and append points belonging to j to cluser of i
-					for (int k=0; k < clusters[j].points().size(); k++) {
-						clusters[i].points().push_back(clusters[j].points()[k]);
-					}
-					//BOOST_LOG_TRIVIAL(debug) << "Merging cluster " << j << " into cluster " << i << ", distance " << diff.norm();
-					clusters.erase(clusters.begin() + j);
-					curr.erase(curr.begin() + j);
-
-					break;
-				}
+	Eigen::VectorXd prevCenter;
+	std::vector<Eigen::VectorXd> currCenters = points;
+
+	// For each point...
+	for(PointIndex i = 0; i < clusters.size(); i++) {
+		prevCenter = points[i];
+		int iterations = 0; //for logging and abort condition
+
+		// ... iterate until convergence
+		do  {
+			iterations++;
+			// perform one iteration of mean shift
+			prevCenter = currCenters[i];
+			currCenters[i] = iterate_mode(currCenters[i], points);
+
+			if (iterations > _max_iterations) {
+				BOOST_LOG_TRIVIAL(warning) << "Max number of iterations for point " << i << " reached - continuing without convergence for this point";
 			}
-		}
+		} while ((prevCenter - currCenters[i]).norm() > _convergence_threshold);
 
-		if (clusters.size() != curr.size())
-			BOOST_LOG_TRIVIAL(warning) << "Cluster count does not equal mode count. (" << clusters.size() << " != " << curr.size() << ")";
+	}
 
-		// Check convergence
-		if (prev.size() == curr.size()) {
-			double total_movement = 0;
-			for (int i=0; i < curr.size(); i++) {
-				total_movement += (curr[i]-prev[i]).norm();
+	// Merge step
+	// For every mode..
+	for (int i=0; i<currCenters.size(); i++) {
+		// Check modes we haven't already executed the (i)-loop for
+		for (int j=i+1; j < currCenters.size(); j++) {
+			Eigen::VectorXd diff = currCenters[i] - currCenters[j];
+			if (diff.norm() < _merge_threshold) {
+				// Merge clusters. Erase one of the modes corresponding to the clusters and append points belonging to j to cluser of i
+				clusters[i].points().insert(clusters[i].points().begin(), clusters[j].points().begin(), clusters[j].points().end());
```

Comment:
Note: with the 2 phase approach, this line can be further simplified to `clusters[i].points().push_back(j);` (we always merge j into i, while j > i, hence j is always a cluster as initialized at the start).
There are probably further clean ups, but I'd push them to our optimization phase.